lib/services/graph-progress.js

```diff
@@ -0,0 +1,232 @@
+
+        return eventsProtocol.publishProgressEvent(
+            graph.instanceId,
+            progress.getProgressEventData()
+        )
+        .catch(function(error) {
+            if(swallowError) {
+                logger.error('Error publishing graph ' + graphState + ' progress event', {
+                    graphId: graph.instanceId,
+                    _status: graph._status,
+                    error: error
+                });
+            } else {
+                throw error;
+            }
+        });
+    };
+
+    /**
+     * Publish a graph started progress event over AMQP and via Web Hooks.
+     *
+     * @param {Object} graph
+     * @param {Boolean} swallowError
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishGraphStarted = function(graph, swallowError) {
+        return this._publishGraphProgress(graph, 'started', swallowError);
+    };
+
+    /**
+     * Publish a graph finished progress event over AMQP and via Web Hooks.
+     *
+     * @param {Object} graph
+     * @param {String} graphState
+     * @param {Boolean} swallowError
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishGraphFinished = function(
+        graph,
+        graphState,
+        swallowError
+    ) {
+        return this._publishGraphProgress(graph, graphState, swallowError);
+    };
+
+    /**
+     * Publishes a task started event over AMQP and via Web Hooks
+     *
+     * @param {Object} task
+     * @param {Boolean} swallowError
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishTaskStarted = function(task, swallowError) {
+        return waterline.graphobjects.findOne({ instanceId: task.context.graphId })
+        .then(function(graph) {
+            var taskFriendlyName = _.get(task, 'definition.friendlyName');
+            var taskId = task.instanceId;
+            var progressMaximum = _.size(_.get(task, 'definition.options.progressMilestones')) ||
+                    Constants.Progress.DefaultTaskProgressMaximum;
+            var taskProgress = {
+                value: 0,
+                maximum: progressMaximum,
+                description: "Task started"
+            };
+
+            var graphDescription = 'Task "' + taskFriendlyName + '" started';
+            var progress = GraphProgress.create(graph, graphDescription);
+            progress.updateTaskProgress(taskId, taskProgress);
+            return eventsProtocol.publishProgressEvent(graph.instanceId,
+                                                       progress.getProgressEventData());
+        })
+        .catch(function(error) {
+            if(swallowError) {
+                logger.error('Error publishing progress event when task started', {
+                    taskId: task.instanceId,
+                    graphId: task.context.graphId,
```

Comment:
Image the `task` has a null context, `task.context.graphId` will throw exception, but what you are doing here is swallowing   exception. The swallow error code itself throw an error! Add some assertion for the input argument at the beginning will help to get rid of such potential error.